src/action-creators/updateThoughtsFromSubscription.ts

```diff
@@ -0,0 +1,46 @@
+import _ from 'lodash'
+import { updateThoughts } from '../action-creators'
+import { ThoughtUpdates, Thunk } from '../@types'
+import { SessionType, getSessionId, getSessionType } from '../util/sessionManager'
+
+interface Updateable {
+  updatedBy?: string
+}
+
+/** Returns true if the updateable entity is coming from the given SessionType and is not self. */
+export const isValidSource = (updateable: Updateable, updateType: SessionType) => {
+  // on a change event, the firebase snapshot only contains updated fields, which may not include updatedBy
+  // in this case, return true so it gets added locally
+  return (
+    !updateable.updatedBy ||
+    (updateable.updatedBy !== getSessionId() && getSessionType(updateable.updatedBy) === updateType)
```

Comment:
Currently, a tab session id is added to the local session list only after visibility has changed. In the case where we open two tabs with a window split or in a dual-screen setup, the visibility of the two tabs doesn't change. So updates from local DB are rejected because they are mistaken for remote updates. Maybe we should call `keepAlive` once at the initialization.

Reply:
Good idea. I may use a dedicated init function, which seems like a better place to initialize the in-memory `sessionId` too.

Reply:
The dedicated `init` won't work without some larger refactoring, because `getSessionId` is used in `initialState`, which gets called before `initialize`. I will just call `keepalive` in `initialize` and otherwise keep the structure as-is.

Reply:
Should we call `keepalive` also on edit to prevent the session from expiring if the tab stays visible?

Reply:
I added a simple middleware that calls `keepalive` on any action.